src/shogun/base/subset_iterators.h

```diff
@@ -0,0 +1,191 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_SUBSET_ITERATORS_H
+#define SHOGUN_SUBSET_ITERATORS_H
+
+#include <shogun/lib/SGVector.h>
+#include <shogun/mathematics/Math.h>
+
+namespace shogun
+{
+	template <typename T>
+	struct remove_cvptr
+	{
+		typedef std::remove_cv_t<std::remove_pointer_t<T>> type;
+	};
+
+	template <typename T>
+	using remove_cvptr_t = typename remove_cvptr<T>::type;
+
+	template <typename T>
+	class CDenseFeatures;
+
+	class CFeatures;
+	class CLabels;
+
+	template <typename T>
+	struct return_type_subset_iterator
+	{
+	};
+
+	template <typename T>
+	struct return_type_subset_iterator<CDenseFeatures<T>>
+	{
+		using type = SGVector<T>;
+	};
+
+	template <template <typename> class IterableSubsetContainer, typename ST>
+	class SubsetIteratorBase
+	{
+	public:
+		template <
+		    typename T, bool is_const = false,
+		    std::enable_if<
+		        (std::is_base_of_v<T, CFeatures>) ||
+		        (std::is_base_of_v<T, CLabels>)>* = nullptr>
+		class subset_iterator
+		{
+		public:
+			using iterator_category = std::forward_iterator_tag;
+			using value_type = typename std::conditional_t<
+			    is_const,
+			    const typename return_type_subset_iterator<
+			        remove_cvptr_t<T>>::type,
+			    typename return_type_subset_iterator<remove_cvptr_t<T>>::type>;
+			using difference_type = index_t;
+			// not a reference but is used by stl algorithms to check type
+			using reference = typename std::conditional_t<
+			    is_const,
+			    const typename return_type_subset_iterator<
+			        remove_cvptr_t<T>>::type,
+			    typename return_type_subset_iterator<remove_cvptr_t<T>>::type>;
+			using pointer = typename std::conditional_t<
+			    is_const,
+			    const typename return_type_subset_iterator<
+			        remove_cvptr_t<T>>::type*,
+			    typename return_type_subset_iterator<remove_cvptr_t<T>>::type*>;
+
+			using internal_pointer = typename std::conditional_t<
+			    is_const, const remove_cvptr_t<T>*, remove_cvptr_t<T>*>;
```

Comment:
the logic is a bit convoluted in the whole class but i wanted to avoid writing two class: one for iterator and one for const_iterator

Reply:
yep, looks ok to me for now